### ISSUE TITLE
Center of pixel option not working

### ISSUE THREAD

**Issue by rosshadden** [kylemacfarlane/zplgrf#6]
I can't get the `center_of_pixel` option to work. I get the following error:

```
Error: /undefined in .setfilladjust
Operand stack:
   true   0
Execution stack:
   %interp_exit   .runexec2   --nostringval--   --nostringval--   --nostringval--   2   %stopped_push   --nostringval--   --nostringval--   --nostringval--   false   1   %stopped_push   .runexec2   --nostringval--   --nostringval--   --nostringval--   2   %stopped_push   --nostringval--
Dictionary stack:
   --dict:729/1123(ro)(G)--   --dict:0/20(G)--   --dict:75/200(L)--
Current allocation mode is local
Last OS error: No such file or directory
GPL Ghostscript 9.27: Unrecoverable error, exit code 1
```

This is happening when I run the underlying generated ghostscript command manually as well, so I don't know if it's a problem with the command being generated or ghostscript itself. Or my system somehow. I'm hoping the former because I have hardly found anything in ghostscript about the `setfilladjust` option and do not feel strongly about being able to get help with it.

The only option I am passing is `center_of_pixel`, though I have tried `use_bindings` as well (and installing the ghostscript python module). The command being generated is:

```
gs -dQUIET -dPARANOIDSAFER -dNOPAUSE -dBATCH -dNOINTERPOLATE -sDEVICE=pngmono -dAdvanceDistance=1000 -r203 -dDEVICEWIDTHPOINTS=288 -dDEVICEHEIGHTPOINTS=432 -dFIXEDMEDIA -dPDFFitPage -sstdout=%stderr -sOutputFile=%stdout -c \<\</.HWMargins\[0.000000 0.000000 0.000000 0.000000\] /Margins\[0 0\]\>\>setpagedevice 0 .setfilladjust -f -
```

My code in case it helps:

```python
from zplgrf.src.zplgrf import GRF
import sys

with open("./file.pdf", "rb") as pdf:
    try:
        pages = GRF.from_pdf(pdf.read(), "plzhelp", center_of_pixel=True)
    except Exception as ex:
        print(ex)
        sys.exit(0)

for grf in pages:
    print(grf.to_zpl())
```

Edit: I should note that it works very well without this option. The result is just blurry as you said it might be in your readme.

**Comment by kylemacfarlane** [comment]
I would guess something is wrong with the PDF. Can you upload it anywhere?

`0 .setfilladjust` is the centre of pixel command. It makes barcodes better but degrades text.

https://github.com/OpenPrinting/cups-filters/blob/master/filter/gstoraster.c#L898

It seems to be completely undocumented. Even the Ghostscript devs didn't know about it when I asked them about improving barcodes. Some random Chinese guy discovered it and showed up four months later:

https://bugs.linuxfoundation.org/show_bug.cgi?id=1373

The only documentation I could find was mention in an ancient 2001 Ghostscript help file still hosted on MIT's website:

http://web.mit.edu/ghostscript/www/Language.htm

**Comment by rawsp33d** [comment]
Thank you for the quick and informative response! Here is an example PDF with which I have this problem: https://cl.ly/a30203c348b7

I don't think there is an issue with the PDF, but I could be wrong. It works fine without the pixel center option. I found that the barcodes are only blurry if viewed at a size under some threshold. If I make the image bigger the barcodes can be picked up by barcode readers. It's possible this is good enough for me but I really would like to get the `center_of_pixel` option working if possible.

**Comment by kylemacfarlane** [comment]
It works for me up to Ghostscript 9.26.

This commit from February mentions the filladjust options: http://git.ghostscript.com/?p=ghostpdl.git;a=commitdiff;h=25c744f8a7761676d7790f513ca138eaaed2c366

So in 9.27 onwards it's only available as .setfilladjust2 as originally documented back in 2001. It's also good to hear that they use it themselves and plan to redocument it as I was worried they'd remove it.

For now either downgrade to 9.26 or I think change this line:
https://github.com/kylemacfarlane/zplgrf/blob/4add6be056aa0da917894f6e4fa5fac499e923cf/src/zplgrf/__init__.py#L435

To `cmd += ['0 0 .setfilladjust2']` (notice it's two zeroes now, I'm only assuming 0 0 is the right equivalent for the old value).

I'll update the package later.

**Comment by kylemacfarlane** [comment]
I updated with a fix and the tests pass against 9.27 but 9.22-9.26 are no longer supported.

It turns out that .setfilladjust2 wasn't available between 9.22-9.26 and I don't want to add overhead of version sniffing. Many distros come with ancient versions of gs that need upgrading already so it shouldn't cause much fuss to set the bar even higher.

**Comment by rawsp33d** [comment]
Your update worked great, thank you.
